utils/py/model_catalog.py

```
@@ -0,0 +1,336 @@
+        return problems
+
+    # Neither commit field can be verified offline (the sha256s bind the actual bytes; these name
+    # where the bytes came from). A format assert is the part that IS checkable here, and it is
+    # what catches a typo'd or truncated provenance reference that would otherwise pass forever.
+    # Full 40-hex only: an abbreviation can grow ambiguous as the upstream repo gains commits.
+    for key in ("tag_commit", "renderer_commit"):
+        if not re.fullmatch(r"[0-9a-f]{40}", str(pin[key])):
+            problems.append(
+                f"{PIN_REL}: {key}={pin[key]!r} is not a full 40-character commit sha — a pin must "
+                f"name an immutable, unambiguous {UPSTREAM_REPO} commit"
+            )
```

Ruff (0.16.3)

[warning] Use `list.extend` to create a transformed list

(PERF401)